File: src/git.ts
import { promisify } from 'util';
import { execFile } from 'child_process';
import * as path from 'path';
import type { FormatConfig } from './config';
import { normalizeRemoteUrl } from './git-url';

const execFileAsync = promisify(execFile);

export type Result<T> = { ok: true; value: T } | { ok: false; error: string };

export interface GitInfo {
  repoRoot: string;
  remoteUrl: string; // normalized HTTPS, no .git suffix
  host: string;
  ref: string;
  pathFromRoot: string;
}

async function git(args: string[], cwd: string): Promise<Result<string>> {
  try {
    const { stdout } = await execFileAsync('git', args, { cwd });
    return { ok: true, value: stdout.trim() };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return { ok: false, error: message };
  }
}

export async function getGitInfo(
  filePath: string,
  cfg: FormatConfig,
): Promise<Result<GitInfo>> {
  const fileDir = path.dirname(filePath);

  const root = await git(['rev-parse', '--show-toplevel'], fileDir);
  if (!root.ok) {
    return { ok: false, error: 'Not in a git repository' };
  }
  const repoRoot = root.value;

  const remote = await git(
    ['config', '--get', `remote.${cfg.gitRemote}.url`],
    repoRoot,
  );
  if (!remote.ok || !remote.value) {
    return { ok: false, error: `Remote '${cfg.gitRemote}' not found` };
  }
  const normalized = normalizeRemoteUrl(remote.value);
  if (!normalized) {
    return { ok: false, error: `Could not parse remote URL: ${remote.value}` };
  }
  if (normalized.host !== 'github.com') {
    return { ok: false, error: `Unsupported git host: ${normalized.host}` };
  }

  let ref: string;
  if (cfg.gitRef === 'branch') {
    const branch = await git(['symbolic-ref', '--short', 'HEAD'], repoRoot);
    if (branch.ok) {
      ref = branch.value;
    } else {
      // Detached HEAD: fall back to short SHA
      const shortSha = await git(['rev-parse', '--short', 'HEAD'], repoRoot);
      if (!shortSha.ok) {
        return { ok: false, error: 'No commits in repository' };
      }
      ref = shortSha.value;
    }
  } else {
    const sha = await git(['rev-parse', 'HEAD'], repoRoot);
    if (!sha.ok) {
      return { ok: false, error: 'No commits in repository' };
    }
    ref = sha.value;
  }

  // pathFromRoot uses POSIX separators (URL form), even on Windows.
  const pathFromRoot = path
    .relative(repoRoot, filePath)
    .split(path.sep)
    .join('/');

  return {
    ok: true,
    value: {
      repoRoot,
      remoteUrl: normalized.httpsBase,
      host: normalized.host,
      ref,
      pathFromRoot,
    },
  };
}

File: nvim/lua/copy-path-line/git.lua
local url_mod = require('copy-path-line.url')

local M = {}

local function run(cmd, cwd)
  local ok, sysmod = pcall(function() return vim.system end)
  if ok and sysmod and type(sysmod) == 'function' then
    local res = vim.system(cmd, { cwd = cwd, text = true }):wait()
    if res.code ~= 0 then
      return nil, (res.stderr or ''):gsub('%s+$', '')
    end
    return (res.stdout or ''):gsub('%s+$', ''), nil
  end
  -- Fallback for Neovim < 0.10
  local out = vim.fn.systemlist(table.concat(cmd, ' ') .. ' 2>/dev/null')
  if vim.v.shell_error ~= 0 then
    return nil, 'git command failed'
  end
  return table.concat(out, '\n'):gsub('%s+$', ''), nil
end

function M.get_info(file_path, cfg)
  local file_dir = vim.fn.fnamemodify(file_path, ':h')

  local root, err = run({ 'git', '-C', file_dir, 'rev-parse', '--show-toplevel' })
  if not root or root == '' then
    return nil, 'Not in a git repository'
  end

  local remote = run({ 'git', '-C', root, 'config', '--get', 'remote.' .. cfg.git_remote .. '.url' })
  if not remote or remote == '' then
    return nil, "Remote '" .. cfg.git_remote .. "' not found"
  end

  local normalized = url_mod.normalize_remote_url(remote)
  if not normalized then
    return nil, 'Could not parse remote URL: ' .. remote
  end
  if normalized.host ~= 'github.com' then
    return nil, 'Unsupported git host: ' .. normalized.host
  end

  local ref
  if cfg.git_ref == 'branch' then
    local branch = run({ 'git', '-C', root, 'symbolic-ref', '--short', 'HEAD' })
    if branch and branch ~= '' then
      ref = branch
    else
      local short_sha = run({ 'git', '-C', root, 'rev-parse', '--short', 'HEAD' })
      if not short_sha or short_sha == '' then
        return nil, 'No commits in repository'
      end
      ref = short_sha
    end
  else
    local sha = run({ 'git', '-C', root, 'rev-parse', 'HEAD' })
    if not sha or sha == '' then
      return nil, 'No commits in repository'
    end
    ref = sha
  end

  -- POSIX-style relative path
  local rel = vim.fn.fnamemodify(file_path, ':p')
  rel = rel:sub(#root + 2) -- strip "<root>/"

  return {
    repo_root = root,
    remote_url = normalized.https_base,
    host = normalized.host,
    ref = ref,
    path_from_root = rel,
  }, nil
end

return M
